x-pack/plugins/security/public/management/edit_role/edit_role_page.tsx

```diff
@@ -17,147 +18,283 @@ import {
+          }
 
-  public render() {
-    const description = this.props.spacesEnabled ? (
-      <FormattedMessage
-        id="xpack.security.management.editRole.setPrivilegesToKibanaSpacesDescription"
-        defaultMessage="Set privileges on your Elasticsearch data and control access to your Kibana spaces."
-      />
-    ) : (
-      <FormattedMessage
-        id="xpack.security.management.editRole.setPrivilegesToKibanaDescription"
-        defaultMessage="Set privileges on your Elasticsearch data and control access to Kibana."
-      />
-    );
+          fetchedRole.elasticsearch.indices.push(emptyOption);
+        }
 
-    return (
-      <div className="editRolePage">
-        <EuiForm {...this.state.formError}>
-          {this.getFormTitle()}
+        setRole(action === 'clone' ? prepareRoleClone(fetchedRole) : copyRole(fetchedRole));
+      })
+      .catch((err: IHttpFetchError) => {
+        if (err.response?.status === 404) {
+          notifications.toasts.addDanger({
+            title: i18n.translate('xpack.security.management.roles.roleNotFound', {
+              defaultMessage: 'No "{roleName}" role found.',
+              values: { roleName },
+            }),
+          });
+          backToRoleList();
+        } else {
+          fatalErrors.add(err);
+        }
+      });
+  }, [roleName, action, fatalErrors, rolesAPIClient, notifications, license]);
 
-          <EuiSpacer />
+  return [role, setRole] as [Role | null, typeof setRole];
+}
 
-          <EuiText size="s">{description}</EuiText>
+function useSpaces(http: HttpStart, fatalErrors: FatalErrorsSetup, spacesEnabled: boolean) {
+  const [spaces, setSpaces] = useState<Space[] | null>(null);
+  useEffect(() => {
+    (spacesEnabled ? http.get('/api/spaces/space') : Promise.resolve([])).then(
+      fetchedSpaces => setSpaces(fetchedSpaces),
+      err => fatalErrors.add(err)
+    );
+  }, [http, fatalErrors, spacesEnabled]);
 
-          {isReservedRole(this.state.role) && (
-            <Fragment>
-              <EuiSpacer size="s" />
-              <EuiText size="s" color="subdued">
-                <p id="reservedRoleDescription" tabIndex={0}>
-                  <FormattedMessage
-                    id="xpack.security.management.editRole.modifyingReversedRolesDescription"
-                    defaultMessage="Reserved roles are built-in and cannot be removed or modified."
-                  />
-                </p>
-              </EuiText>
-            </Fragment>
-          )}
+  return spaces;
+}
 
-          <EuiSpacer />
+function useFeatures(http: HttpStart, fatalErrors: FatalErrorsSetup) {
+  const [features, setFeatures] = useState<Feature[] | null>(null);
+  useEffect(() => {
+    http.get('/api/features').then(
+      fetchedFeatures => setFeatures(fetchedFeatures),
+      (err: IHttpFetchError) => {
+        // TODO: This check can be removed once all of these `resolve` entries are moved out of Angular and into the React app.
```

Comment:
__note:__ honestly don't know what that means, just copied from the LP.

Reply:
Currently, the `/api/features` endpoint effectively requires the "Global All" kibana privilege (e.g., what the `kibana_user` grants), because it returns information about all registered features (https://github.com/elastic/kibana/pull/35841).

It's possible that a user with `manage_security` will attempt to visit the role management page without the correct Kibana privileges. If that's the case, then they receive a partial view of the role, and the UI does not allow them to make changes to that role's kibana privileges. When this user visits the edit role page, this API endpoint will throw a `404`, which causes this `resolve` entry to fail. This failure silently prevents the screen from loading _at all_, so we instead attempt to detect the `404` here, and respond in a way that still allows the UI to render itself.

Reply:
Aha, thanks a lot for explaining! I'll add that as a comment here.